### DISCUSSION TITLE
Value Objects and Embeddables

### DISCUSSION THREAD

**Discussion by zkulbeda** [mikro-orm/mikro-orm#3685] [Ideas]
Hi, I'm studying DDD. Is there some the best practies how to store readonly value objects in MikroORM?

I found that embeddables are the best choice right now for this, but now they can't offer functionality as full as properties on entities, for example, I can't use the `@PrimaryKey` directive on embeddable's property. I believe this restricted because depending on `@Embeddable` options, the embeddable can become a json object in database.

```typescript
@Embeddable()
export class UserId {
  @PrimaryKey()
  public readonly id: string;

  public constructor(id: string) {
    if(id.length < 0 && id.length > 32) throw new Error("bad id");
    this.id = id;
  }
}
```

Also I considered custom types in MikroORM, but then desided it much more complicated to create both custom type and its value object for every type of value object.

See my full usage example:
```typescript
@Embeddable()
export class PasswordHash {
  @Property()
  public readonly hash: string;

  @Property()
  public readonly salt: string;

  public constructor(values: { hash: string; salt: string }) {
    const result = PasswordHashSchema.safeParse(values);
    if (!result.success)
      throw new Error("Invalid PasswordHash", { cause: result.error });
    this.hash = values.hash;
    this.salt = values.salt;
  }
}

@Embeddable()
export class Username{
  @Unique()
  @Property()
  public readonly username: string;

  public constructor(value: string) {
    const result = zod.string().min(4).max(128).regex(/^(?:[^\.\-])[\w\.-]+(?:[^\.\-])$/).safeParse(value);
    if (!result.success)
      throw new Error("Invalid Username", { cause: result.error });
    this.username = value;
  }
}

@Entity()
export class User {
  @Embedded(() => UserId, { prefix: false })
  public userId!: UserId = new UserId(cuid());

  @Embedded(() => PasswordHash, { prefix: false })
  public passwordHash!: PasswordHash;

  @Embedded(() => Username, { prefix: false })
  public username!: PasswordHash;

  // ideally these fields should have private setter, which typescript can't provide

  // ... entity consists of only embedded fields and some methods
}
```
So, all embeddables will become readonly value objects that have custom validation inside. So retrieving and persisting an Aggregate Root will be as simple as calling `find` from `EntityManager`. Right now I'm writing a wrapper for regular entities, that repeats behavior of a real aggregate.

**Comment by B4nan** [comment]
> the embeddable can become a json object in database.

Moreover, you can also nest embeddables, and that works in the default inline mode as well. So the reason is simply to reduce confusion, as it would work only in very simple conditions (inline mode for one level), and still could cause confusion (using multiple embeddables with PK -> entity with composite PK).

I don't have any guidance around this, personally I'd prefer custom mapped types over a single property embeddables as it will result in simpler hydration and will be probably faster perf-wise.
